### PR TITLE
[pten] Cast xpu kernel

### PR DESCRIPTION
**Pull request by MingMingShangTian**
### PR types

Others
### PR changes

Others
### Describe

Add Cast XPU device kerenl.

### CODE REVIEW

File: paddle/pten/kernels/xpu/cast_kernel.cc [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "paddle/pten/kernels/cast_kernel.h"
+
+#include "paddle/fluid/platform/device/xpu/xpu_header.h"
+#include "paddle/pten/backends/xpu/xpu_context.h"
+#include "paddle/pten/core/kernel_registry.h"
+
+#include "paddle/fluid/platform/enforce.h"
```
Comment (chenwhql) [inline paddle/pten/kernels/xpu/cast_kernel.cc]: include pten/core/enforce.h

File: paddle/pten/kernels/xpu/cast_kernel.cc [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+
+namespace pten {
+
+template <typename T, typename Context>
+void CastKernel(const Context& dev_ctx,
+                const DenseTensor& x,
+                DataType out_dtype,
+                DenseTensor* out) {
+  using XPUInTDType = typename XPUTypeTrait<T>::Type;
+  using float16 = typename XPUTypeTrait<paddle::platform::float16>::Type;
```
Comment (chenwhql) [inline paddle/pten/kernels/xpu/cast_kernel.cc]: pten::dtype::float16

File: paddle/pten/kernels/xpu/cast_kernel.cc [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+          reinterpret_cast<const XPUInTDType*>(in_data),
+          out->mutable_data<float>(dev_ctx.GetPlace()),
+          numel);
+      break;
+    case pten::DataType::FLOAT16:
+      r = xpu::cast_v2<XPUInTDType, float16>(
+          dev_ctx.x_context(),
+          reinterpret_cast<const XPUInTDType*>(in_data),
+          reinterpret_cast<float16*>(
+              out->mutable_data<paddle::platform::float16>(dev_ctx.GetPlace())),
```
Comment (chenwhql) [inline paddle/pten/kernels/xpu/cast_kernel.cc]: 同上

File: paddle/pten/kernels/xpu/cast_kernel.cc [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+}
+}  // namespace pten
+
+PT_REGISTER_KERNEL(cast,
+                   XPU,
+                   ALL_LAYOUT,
+                   pten::CastKernel,
+                   int32_t,
+                   float,
+                   paddle::platform::float16,
```
Comment (chenwhql) [inline paddle/pten/kernels/xpu/cast_kernel.cc]: 同上

File: paddle/pten/kernels/xpu/cast_kernel.cc [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+      break;
+    default:
+      PADDLE_THROW(platform::errors::Unavailable(
+          "Not supported cast %d -> %d", x.dtype(), out_dtype));
+  }
+
+  PADDLE_ENFORCE_EQ(
+      r,
+      XPU_SUCCESS,
+      platform::errors::External(
```
Comment (chenwhql) [inline paddle/pten/kernels/xpu/cast_kernel.cc]: pten::errors

File: paddle/pten/kernels/xpu/cast_kernel.cc (hunk 1 above)
Comment (MingMingShangTian) [reply paddle/pten/kernels/xpu/cast_kernel.cc]: Done, thanks.

File: paddle/pten/kernels/xpu/cast_kernel.cc (hunk 2 above)
Comment (MingMingShangTian) [reply paddle/pten/kernels/xpu/cast_kernel.cc]: Done

File: paddle/pten/kernels/xpu/cast_kernel.cc (hunk 3 above)
Comment (MingMingShangTian) [reply paddle/pten/kernels/xpu/cast_kernel.cc]: Done

File: paddle/pten/kernels/xpu/cast_kernel.cc (hunk 4 above)
Comment (MingMingShangTian) [reply paddle/pten/kernels/xpu/cast_kernel.cc]: Done

File: paddle/pten/kernels/xpu/cast_kernel.cc (hunk 5 above)
Comment (MingMingShangTian) [reply paddle/pten/kernels/xpu/cast_kernel.cc]: Done

**Review by chenwhql — APPROVED** [review]

**Pull request merged**